### PR TITLE
Adopt etcd resources

### PR DESCRIPTION
**Pull request by georgekuruvillak**
`etcd-druid` controller used create the resources such as `StatefulSet`, `Service` and `Configmap` when it acted on an `Etcd` resource. However, there will be scenarios where it will have to adopt existing resources (deployed prior to adoption to `etcd-druid`. This PR implements the logic which adopts exisitng `etcd` related resources.

### CODE REVIEW

File: config/manager/manager.yaml [hunk 1]
```diff
@@ -2,33 +2,33 @@ apiVersion: v1
...
       containers:
       - command:
-        - /manager
+        - /bin/druid
         args:
         - --enable-leader-election
-        image: controller:latest
-        name: manager
+        name: druid
+        imagePullPolicy: Always
```
Comment (amshuman-kr) [inline config/manager/manager.yaml]: Should we use the `Always` policy?

File: config/samples/druid_v1alpha1_etcd.yaml [hunk 2]
```diff
@@ -33,7 +32,7 @@ spec:
         clientPort: 2379
         serverPort: 2380
     backup:
-        version: 0.8.0-dev
+        version: 0.8.0-dev-dd26b65a2ab4efd77805584b33e8c559dd3c6d53
```
Comment (amshuman-kr) [inline config/samples/druid_v1alpha1_etcd.yaml]: Is this change required?

File: controllers/etcd_controller.go [hunk 3]
```diff
@@ -159,30 +179,159 @@ func (r *EtcdReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
-	images, err := imagevector.FindImages(imageVector, []string{"etcd", "etcd-backup-restore"})
-	if err != nil {
+	if err := r.updateEtcdStatus(etcd); err != nil {
 		return ctrl.Result{
 			Requeue:      true,
 			RequeueAfter: time.Second * 5,
 		}, err
 	}
 
+	logger.Infof("Succesfully reconciled etcd: %s", etcd.GetName())
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]:
```suggestion
	logger.Infof("Successfully reconciled etcd: %s", etcd.GetName())
```

File: controllers/etcd_controller.go [hunk 4]
```diff
@@ -159,30 +179,159 @@ func (r *EtcdReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+
+	return ctrl.Result{
+		Requeue:      true,
+		RequeueAfter: time.Second * 5,
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]: Should the requeue rate be the same for success as well as error?

File: controllers/etcd_controller.go [hunk 5]
```diff
@@ -159,30 +179,159 @@ func (r *EtcdReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+		return false, err
+	}
+
+	etcdConfigMapList, err := r.getEtcdConfigMap(etcd)
+	if err != nil {
+		return false, err
+	}
+
+	if (len(etcdServiceList.Items) + len(etcdStatefulsetList.Items) + len(etcdConfigMapList.Items)) == 0 {
+		return false, nil
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]: Can we log a message here which resources were found or not found?

File: controllers/etcd_controller.go [hunk 6]
```diff
@@ -159,30 +179,159 @@ func (r *EtcdReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+			if err != nil {
+				return true, err
+			}
+		}
+	}
+
+	return true, nil
+}
+
+func checkForEtcdOwnerReference(refs []metav1.OwnerReference, etcd *druidv1alpha1.Etcd) bool {
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]: This checks only if any of the refs have the UID of etcd. This is necessary but [not sufficient](https://stupefied-goodall-e282f7.netlify.com/contributors/design-proposals/api-machinery/controller-ref/#adoption). We should not adopt objects with refs to some other controller.

It would be better to rename this func as `getAdoptionState(refs []metav1.OwnerReference, etcd *druidv1alpha1.Etcd) adoptionState` with values `ADOPTED`, `CAN_ADOPT`, `CANNOT_ADOPT`.

File: controllers/etcd_controller.go [hunk 7]
```diff
@@ -390,14 +530,14 @@ func (r *EtcdReconciler) updateStatusFromServices(etcd *druidv1alpha1.Etcd) erro
...
+func (r *EtcdReconciler) updateEtcdErrorStatus(etcd, etcdCopy *druidv1alpha1.Etcd, lastError error) error {
 
-	etcdStatus := etcd.Status
+	etcdStatus := etcdCopy.Status
 	lastErrStr := fmt.Sprintf("%v", lastError)
 	etcdStatus.LastError = &lastErrStr
 
-	etcd.Status = etcdStatus
-	if err := r.Status().Update(context.TODO(), etcd); err != nil {
+	etcdCopy.Status = etcdStatus
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]: Why this required?

File: config/samples/druid_v1alpha1_etcd.yaml (hunk 2 above)
Comment (shreyas-s-rao) [reply config/samples/druid_v1alpha1_etcd.yaml]: Until we make a release of etcd-backup-restore, we need to use this image because it contains API changes to etcdbrctl.

Comment (georgekuruvillak) [reply config/samples/druid_v1alpha1_etcd.yaml]: As Shreyas said, we need this till the changes are released.

File: controllers/etcd_controller.go (hunk 5 above)
Comment (georgekuruvillak) [reply controllers/etcd_controller.go]: This logic not used anymore.

File: controllers/etcd_controller.go (hunk 6 above)
Comment (georgekuruvillak) [reply controllers/etcd_controller.go]: I have changed the implementation logic of adoption. PTAL.

File: controllers/etcd_controller.go (hunk 7 above)
Comment (georgekuruvillak) [reply controllers/etcd_controller.go]: Changed.

File: controllers/etcd_controller.go (hunk 4 above)
Comment (georgekuruvillak) [reply controllers/etcd_controller.go]: Changed it to 5 minutes.

File: controllers/etcd_controller.go [hunk 8]
```diff
@@ -159,30 +197,501 @@ func (r *EtcdReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+
+func (r *EtcdReconciler) getServiceFromEtcd(etcd *druidv1alpha1.Etcd) (*corev1.Service, error) {
+	var err error
+	decoded := &corev1.Service{}
+	values, err := r.getMapFromEtcd(etcd)
+	if err != nil {
+		return nil, err
+	}
+	chartPath := getChartPath()
+	renderer, err := r.ChartApplier.Render(chartPath, etcd.Name, etcd.Namespace, values)
```
[the remark this answers is not in the document]
Comment (georgekuruvillak) [reply controllers/etcd_controller.go]: Ah yes. That makes sense.

File: controllers/etcd_controller.go [hunk 9]
```diff
@@ -159,30 +197,501 @@ func (r *EtcdReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+		return nil, err
+	}
+
+	if reflect.DeepEqual(ss.Spec, decoded.Spec) {
+		return ss, nil
+	}
+	ssCopy := ss.DeepCopy()
+	decoded.Spec.DeepCopyInto(&ssCopy.Spec)
+	// Copy ClusterIP as the field is immutable
+	ssCopy.Spec.ClusterIP = ss.Spec.ClusterIP
```
[the remark this answers is not in the document]
Comment (georgekuruvillak) [reply controllers/etcd_controller.go]: I'll error out if the type is changed. In case of ClusterIP, we need to sync the field as it is not in the service chart template. As for type: ClusterIP, we are explicitly setting it in the template. Ill make a note there to refrain from changing the type casually. WDYT?

[the remark this answers is not in the document]
Comment (amshuman-kr) [reply controllers/etcd_controller.go]: 👍 Ok. We will deal with that problem when it is required.

File: Dockerfile [hunk 10]
```diff
@@ -1,25 +1,20 @@
...
 # Refer to https://github.com/GoogleContainerTools/distroless for more details
-FROM gcr.io/distroless/static:latest
+FROM alpine:3.10.3 AS druid
+RUN apk add --update bash curl
 WORKDIR /
-COPY --from=builder /workspace/manager .
-ENTRYPOINT ["/manager"]
+COPY --from=builder /go/src/github.com/gardener/etcd-druid/bin/linux-amd64/etcd-druid bin/.
+COPY charts charts
+ENTRYPOINT ["/bin/etcd-druid --kubeconfig=seci308301-kubeconfig"]
```
[the remark this answers is not in the document]
Comment (georgekuruvillak) [reply Dockerfile]: Oops. That crawled in as part of the test.

File: .ci/check [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+# Install Golint (linting tool).
+GO111MODULE=on go get -u golang.org/x/lint/golint
+
+###############################################################################
+
+PACKAGES="$(GO111MODULE=on go list -mod=vendor -e ./... | grep -vE '/api|/api/v1alpha1/|/tmp/|/vendor/')"
+LINT_FOLDERS="$(echo ${PACKAGES} | sed "s|github.com/gardener/etcd-druid|.|g")"
+
+# Execute static code checks.
```
[the remark this answers is not in the document]
Comment (georgekuruvillak) [reply .ci/check]: exported globally

File: .ci/test [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+  go tool cover -func ${output_dir}/${coverprofile_file}
+}
+
+################################################################################
+
+TEST_PACKAGES="api controllers"
+GINKGO_COMMON_FLAGS="-r -timeout=1h0m0s --randomizeAllSpecs --randomizeSuites --failOnPending --progress"
+if [ -z $COVER ] || [ "$COVER" = false ] ; then
+  echo "[INFO] Test coverage is disabled."
+  ginkgo -race -trace $GINKGO_COMMON_FLAGS ${TEST_PACKAGES}
```
[the remark this answers is not in the document]
Comment (georgekuruvillak) [reply .ci/test]: exported globally

File: .ci/test [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+GO111MODULE=on go get -u github.com/onsi/ginkgo/ginkgo
+
+###############################################################################
+
+function test_with_coverage() {
+  echo "[INFO] Test coverage is enabled."
+  local output_dir=test/output
+  local coverprofile_file=coverprofile.out
+  mkdir -p test/output
+  ginkgo $GINKGO_COMMON_FLAGS --coverprofile ${coverprofile_file} -covermode=set -outputdir ${output_dir} ${TEST_PACKAGES}
```
[the remark this answers is not in the document]
Comment (georgekuruvillak) [reply .ci/test]: exported globally
